crates/matrix-sdk/src/widget/client/mod.rs

```diff
@@ -77,12 +78,37 @@ pub(super) async fn run<T: PermissionsProvider>(
                 }
             },
             Err(e) => {
-                // TODO: We need to construct a message to inform the widget
-                // about the incorrectly formatted message, but it does not look
-                // like we have a well-defined error mechanism for such cases.
-                // Clarify that later.
-                warn!("Failed to parse message from widget: {e}");
+                // Parse the message as a BadMessage
+                let mut issues: Vec<String> = vec![];
+                match from_json::<BadMessage>(&raw) {
+                    Ok(bad_msg) => issues.push(format!("{}", bad_msg)),
+                    Err(_) => issues.push("The message could not be parsed. There most likely is a json syntax issue in the sent request.".to_owned()),
+                }
+                issues.push(format!("\nThe message string:\n{}", &raw));
+                issues.push(format!("\nAdditional error description:\n{}", e));
+                let error_message = issues.join("");
+                warn!("Failed to parse message from widget:\n{}", error_message);
+                Err(Error::custom(error_message))
```

Comment:
It seems like the "bad message" (if any is present) is just another message with a known format, right? If so, I think that instead of additionally parsing it here, we just need to extend our `Message`, so that instead of being `struct Message { header, action }`, it becomes `enum Message { Good { header, action }, Bad { ... } }`, though I'm not sure what's the semantics of a `BadMessage`, i.e. why would the widget send a `BadMessage`? What does it signify? I would understand if it was an _error message_, but it's not in this case.

That being said, I see that the `BadMessage` is nothing else, but our current `Message` with most fields being optional or without stronger types (`serde::Value` instead of concrete types). And the only thing we do with them is serialising them to send them back, right?

I think the correct approach would then be to just parse the `Message` the same we do it right now and if the error occurs during parsing, we just send the error response passing the `raw` message back, telling that it was incorrectly formatted. It will be much shorter in terms of code and would essentially do the same. Or do I miss something?

Also, I'm not sure if we really want to use `serde_json::Value` and access values by name `value_msg["response"]`. I imagine such things are useful whe implementing `Serialize` or `Deserialize`, but not in the actual code where we generate responses (we could construct a structure to represent an error if necessary or just inline it with `json!()` if it's small enough and simple).

Reply:
> I think the correct approach would then be to just parse the Message the same we do it right now and if the error occurs during parsing, we just send the error response passing the raw message back, telling that it was incorrectly formatted. It will be much shorter in terms of code and would essentially do the same. Or do I miss something?

This is basically what it does with the additional step that it checks at what level it failed: json syntax error, some fields broken, fields compleatly off but the parsable json.